File: Platform_Computing/Assignment_4_Generate_Users/Link_User.py
import time
from selenium import webdriver
from selenium.webdriver.common.by import By

'''
Returns true or false on whether a keyword exists
Capitalization should not matter (e.g. will detect "Student" on a page as "student" and vice versa)
Will detect super word in page (e.g. will detect "students" on a page as "student")
'''
def findKeyword(driver, keyword)->bool:
    print(driver.page_source.lower()) 
    return keyword.lower() in driver.page_source.lower()

'''
Returns number of elements given tag_name
'''
def countTagElem(driver, tag_name)->int:
    count = 0
    for tags in tag_name:
         count += len(driver.find_elements(By.TAG_NAME, tags))
    return count

'''
Clicks first link
'''
def clickLink(driver, tag_name):
     #find link 
     links = driver.find_elements(By.TAG_NAME, tag_name)
     # only clicks the first link
     for link in links:
          link.click()

'''
Returns via keyword or image based on action
action : KEYWORD, IMAGE, LINK
driver/web driver: driver
totalRewardTime : value to wait on site
reqList (the request list): list of either keyword or element tag
'''
def userAction(action, driver, rewardTime, reqList)->float:
    totalRewardTime = 0
    if action.upper() == "KEYWORD":
     for keyword in reqList: 
          if findKeyword(driver, keyword):
              print("found", keyword)
              totalRewardTime += rewardTime
              time.sleep(rewardTime)
          else:
              print(keyword, "not found")  
    elif action.upper() == "IMAGE":
     numImages = countTagElem(driver, reqList)
     totalRewardTime = rewardTime*numImages
     time.sleep(totalRewardTime) 
    elif action.upper() == "LINK":
     numLinks = countTagElem(driver, reqList)
     totalRewardTime += rewardTime*numLinks 
     time.sleep(rewardTime) 
     time.sleep(totalRewardTime)

     return totalRewardTime

def main():
    #Initialize browser
    driver = webdriver.Chrome()
    #Navigate to your website
    driver.get("http://localhost:3000/")
    rewardTime = 10
    totalRewardTime = 0
    keywords = ["student", "CSUSB"]
    totalRewardTime = userAction("KEYWORD", driver, rewardTime, keywords)
    tag_name = ["img"]
    totalRewardTime += userAction("IMAGE", driver, rewardTime, tag_name)
    link = ["a"]
    totalRewardTime += userAction("LINK", driver, rewardTime, link)

    clickLink(driver)
    driver.quit()

    print("Presence Time:", totalRewardTime)

if __name__ == "__main__":
    main()

File: Platform_Computing/Assignment_4_Generate_Users/Link_User.js
const {By, Builder, Browser} = require('selenium-webdriver');

async function findKeyword(driver, keyword) {
    console.log((await driver.getPageSource()).toLowerCase());
    return (await driver.getPageSource()).toLowerCase().includes(keyword.toLowerCase());
}

async function countTagElem(driver, tag_name) {
    let count = 0;
    for (const tagNames of tag_name) {
        const elements = await driver.findElements(By.css(tagNames));
        count += elements.length;
    }
    return count; 
}

async function clickLink(driver, tag_name) {
    const links = await driver.findElements(By.css(tag_name));
    for (const link of links) {
        await link.click();
    }
}

//action is a string which is either keyword, image or link
async function userAction(action, driver, rewardTime, reqList) {
    let totalRewardTime = 0;
    if (action.toUpperCase() === "KEYWORD") {
        for (const keyword of reqList) {
            if (await findKeyword(driver, keyword)) {
                console.log("found", keyword);
                totalRewardTime += rewardTime;
                await driver.sleep(rewardTime * 1000); //Multiply by 1000 milliseconds since 1 second is equivalent to 1000 miliseconds
            } else {
                console.log(keyword, "not found");
            }
        }
    } else if (action.toUpperCase() === "IMAGE") {
        const numImages = await countTagElem(driver, reqList);
        totalRewardTime = rewardTime * numImages;
        await driver.sleep(totalRewardTime * 1000); //Multiply by 1000 milliseconds since 1 second is equivalent to 1000 miliseconds
    } else if (action.toUpperCase() === "LINK") {
        const numLinks = await countTagElem(driver, reqList);
        totalRewardTime += rewardTime * numLinks;
        await driver.sleep(totalRewardTime * 1000); //Multiply by 1000 milliseconds since 1 second is equivalent to 1000 miliseconds
    }
    return totalRewardTime;
}

async function main() {
    //Initialize browser
    let driver = await new Builder().forBrowser(Browser.CHROME).build();
    try {
        //Navigate to your website
        await driver.get("http://localhost:3000/");

        const rewardTime = 10;
        let totalRewardTime = 0;

        //comment out the 3 lines of code below to get empty arrays to simulate a Presence Time of 0
        //const keywords = []; //Empty array to simulate no keywords found;
        //const tagNames = []; //Empty array to simulate no images found
        //const link = []; //Empty array to simulate no links found
        //the website will exit right away because the Presence Time is 0 if keywords, tagNames, and link are ALL empty arrays

        //uncomment the 3 lines of code below so that keywords, tagnames, and link are NOT empty arrays to output a Presence Time
        //and not a Presence Time of 0
        const keywords = ["student", "CSUSB"];
        const tagNames = ["img"];
        const link = ["a"];

        //commented out 2 lines of below as LINK is the only desired action, not KEYWORD or IMAGE
        //totalRewardTime += await userAction("KEYWORD", driver, rewardTime, keywords);
        //totalRewardTime += await userAction("IMAGE", driver, rewardTime, tagNames);
        totalRewardTime += await userAction("LINK", driver, rewardTime, link);
        await clickLink(driver, "a");
        console.log("Presence Time:", totalRewardTime);
    } finally {
        await driver.quit();
    }
}

main().catch(console.error);